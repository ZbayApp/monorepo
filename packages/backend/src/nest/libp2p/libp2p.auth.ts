import {
  ComponentLogger,
  Connection,
  PeerId,
  PeerStore,
  serviceCapabilities,
  Stream,
  Topology,
} from '@libp2p/interface'
import type { ConnectionManager, IncomingStreamData, Registrar } from '@libp2p/interface-internal'
import * as Auth from '@localfirst/auth'
import { pipe } from 'it-pipe'
import { encode, decode } from 'it-length-prefixed'

import { SigChainService } from '../auth/sigchain.service'
import { createLogger } from '../common/logger'
import { createQuietLogger } from '@quiet/logger'
import { ConnectionParams } from '3rd-party/auth/packages/auth/dist/connection/Connection'
import { Libp2pService } from './libp2p.service'
import { Libp2pEvents } from './libp2p.types'
import { abortableAsyncIterable } from '../common/utils'

export interface Libp2pAuthComponents {
  peerId: PeerId
  peerStore: PeerStore
  registrar: Registrar
  connectionManager: ConnectionManager
  logger: ComponentLogger
}

export interface Libp2pAuthStatus {
  started: boolean
  joining: boolean
}

enum JoinStatus {
  PENDING = 'PENDING',
  JOINING = 'JOINING',
  JOINED = 'JOINED',
  NOT_STARTED = 'NOT_STARTED',
}

const createLFALogger = createQuietLogger('localfirst')

export class Libp2pAuth {
  private readonly protocol: string
  private readonly components: Libp2pAuthComponents
  private sigChainService: SigChainService
  private libp2pService: Libp2pService
  private authConnections: Map<string, Auth.Connection>
  private peerConnections: Map<string, Connection>
  private bufferedConnections: { peerId: PeerId; connection: Connection }[]
  private unblockInterval: NodeJS.Timeout
  private joinStatus: JoinStatus
  private logger: ReturnType<typeof createLogger> = createLogger('libp2p:auth')
  readonly [serviceCapabilities]: string[] = ['@quiet/auth']
  readonly [Symbol.toStringTag]: string = 'lfaAuth'

  constructor(sigChainService: SigChainService, libp2pService: Libp2pService, components: Libp2pAuthComponents) {
    this.protocol = '/local-first-auth/1.0.0'
    this.components = components
    this.sigChainService = sigChainService
    this.libp2pService = libp2pService
    this.authConnections = new Map()
    this.peerConnections = new Map()
    this.bufferedConnections = []

    if (sigChainService.activeChainTeamName == null) {
      this.logger.warn('No active chain found')
      this.joinStatus = JoinStatus.NOT_STARTED
    } else {
      this.logger = this.logger.extend(sigChainService.getActiveChain().localUserContext.user.userName)
      if (sigChainService.getActiveChain()!.team == null) {
        this.joinStatus = JoinStatus.PENDING
      } else {
        this.joinStatus = JoinStatus.JOINED
      }
    }

    this.logger.info('Auth service initialized')
    this.logger.info('sigChainService', sigChainService.activeChainTeamName)

    // Set up a periodic check to process buffered connections
    this.unblockInterval = setInterval(this.unblockConnections, 5_000, this.bufferedConnections)
  }

  private emit(eventName: string, ...args: any[]) {
    this.libp2pService.emit(eventName, ...args)
  }

  // Process any connections that were buffered because we were waiting for a chain
  private async unblockConnections(conns: { peerId: PeerId; connection: Connection }[]) {
    if (this.joinStatus === JoinStatus.NOT_STARTED && this.sigChainService.activeChainTeamName != null) {
      this.logger.info(`Unblocking ${conns.length} connections now that we have an active chain`)
      this.joinStatus = JoinStatus.PENDING
    } else if (this.joinStatus !== JoinStatus.JOINED) {
      return
    }

    this.logger.info(`Unblocking ${conns.length} buffered connections now that we've joined the chain`)
    while (conns.length > 0) {
      const conn = conns.pop()
      if (conn != null) {
        await this.onPeerConnected(conn.peerId, conn.connection)
      }
    }
  }

  async start() {
    this.logger.info('Auth service starting')

    const topology: Topology = {
      onConnect: this.onPeerConnected.bind(this),
      onDisconnect: this.onPeerDisconnected.bind(this),
      notifyOnLimitedConnection: false,
    }

    const registrar = this.components.registrar
    await registrar.register(this.protocol, topology)
    await registrar.handle(this.protocol, this.onIncomingStream.bind(this), {
      runOnLimitedConnection: false,
    })
  }

  async beforeStop() {
    this.logger.info('beforeStop')
  }

  async stop() {
    this.logger.info('stop')

    // Clear the unblock interval
    clearInterval(this.unblockInterval)

    // Close all auth connections
    for (const peerId of this.authConnections.keys()) {
      this.closeAuthConnection(peerId)
    }

    this.logger.info('Libp2pAuth service stopped')
  }

  async afterStop() {
    this.logger.info('afterStop')
    if (this.sigChainService.activeChainTeamName != null) {
      await this.sigChainService.saveChain(this.sigChainService.activeChainTeamName)
    }
  }

  /**
   * Handle an incoming ephemeral stream.
   * Once the stream is processed, it is closed.
   */
  private async onIncomingStream({ stream, connection }: IncomingStreamData) {
    const peerId = connection.remotePeer
    this.logger.info(`Handling incoming ephemeral stream ${connection.id.toString()} from ${peerId.toString()}`)
    const abortController = new AbortController()

    // Process messages from the stream
    this.handleIncomingMessages(peerId, stream, abortController)
      .catch(err => {
        this.logger.error(`Error processing incoming stream from ${peerId.toString()}`, err)
        if (!abortController.signal.aborted) {
          abortController.abort(err)
        }
      })
      .finally(() => {
        stream
          .close()
          .catch(err => this.logger.error(`Error closing incoming ephemeral stream from ${peerId.toString()}`, err))
      })
  }

  /**
   * Process incoming messages by decoding the length-prefixed data and delivering
   * it to the corresponding auth connection.
   */
  private async handleIncomingMessages(peerId: PeerId, stream: Stream, abortController: AbortController) {
    await pipe(
      stream,
      source => decode(source),
      async source => {
        for await (const data of abortableAsyncIterable(source, abortController.signal)) {
          try {
            const authConn = this.authConnections.get(peerId.toString())
            if (!authConn) {
              this.logger.error(`No auth connection established for ${peerId.toString()}`)
            } else {
              authConn.deliver(data.subarray())
            }
          } catch (e) {
            this.logger.error(`Error while delivering message to ${peerId.toString()}`, e)
            if (!abortController.signal.aborted) {
              abortController.abort(e)
            }
          }
        }
      }
    )
  }

  /**
   * Send an outgoing message using an ephemeral stream.
   * This method opens a new stream, writes the encoded message, and then closes it.
   */
  private async sendMessage(peerId: PeerId, message: Uint8Array) {
    const connection = this.peerConnections.get(peerId.toString())
    if (!connection) {
      this.logger.warn(`No connection available for ephemeral stream to ${peerId.toString()}`)
      return
    }

    const abortController = new AbortController()
    try {
      this.logger.info(`Opening ephemeral outbound stream to ${peerId.toString()}`)
      const stream = await connection.newStream(this.protocol, {
        runOnLimitedConnection: false,
        negotiateFully: false,
        signal: abortController.signal,
      })
      this.logger.info(`Ephemeral stream opened to ${peerId.toString()}, sending message`)
      if (stream.status !== 'open') {
        this.logger.warn(
          `Attempted to send message to ${peerId.toString()} on ephemeral stream that had already closed`
        )
        return
      }
      await pipe([encode.single(message)], stream)
      await stream.close()
      this.logger.info(`Ephemeral stream closed to ${peerId.toString()}`)
    } catch (e) {
      this.logger.error(`Error sending ephemeral message to ${peerId.toString()}`, e)
      if (!abortController.signal.aborted) {
        abortController.abort(e)
      }
    }
  }

  /**
   * Called when a peer connects. If weâ€™re not ready to start (e.g. no active chain),
   * the connection is buffered. Otherwise we create a new auth connection and
   * store the underlying libp2p connection for ephemeral stream use.
   */
  private async onPeerConnected(peerId: PeerId, connection: Connection) {
    if (this.authConnections.has(peerId.toString())) {
      this.logger.info(`Auth connection with ${peerId.toString()} already exists`)
      return
    }
    if (this.joinStatus === JoinStatus.JOINING) {
      this.logger.warn(`Connection to ${peerId.toString()} will be buffered due to a concurrent join`)
      this.bufferedConnections.push({ peerId, connection })
      return
    }
    if (this.sigChainService.activeChainTeamName == null) {
      this.logger.warn(`No active chain found, buffering connection to ${peerId.toString()}`)
      this.bufferedConnections.push({ peerId, connection })
      return
    }

    if (this.joinStatus === JoinStatus.PENDING) {
      this.joinStatus = JoinStatus.JOINING
    }

    this.logger.info(`Peer connected (direction = ${connection.direction})! (status = ${connection.status})`)
    if (connection.status !== 'open') {
      this.logger.warn(`The connection with ${peerId.toString()} was not in an open state!`)
      return
    }

    const context = this.sigChainService.getActiveChain().context

    if (this.authConnections.has(peerId.toString())) {
      this.logger.info(
        `A connection with ${peerId.toString()} was already available, skipping connection initialization!`
      )
      return
    }

    // Create an auth connection using an ephemeral sendMessage callback.
    const authConnection = new Auth.Connection({
      context,
      sendMessage: (message: Uint8Array) => {
        // Fire-and-forget: send message using an ephemeral stream.
        this.sendMessage(peerId, message).catch(err => {
          this.logger.error(`Error in sendMessage callback for ${peerId.toString()}`, err)
        })
      },
      createLogger: createLFALogger,
    } as ConnectionParams)

    // Set up auth connection event handlers.
    authConnection.on('connected', () => {
      if (this.sigChainService.activeChainTeamName != null) {
        this.logger.debug(`Sending sync message because our chain is initialized`)
        const sigChain = this.sigChainService.getActiveChain()
        const team = sigChain.team
        const user = sigChain.localUserContext.user
        authConnection.emit('sync', { team, user })
      }
      this.emit(Libp2pEvents.AUTH_CONNECTED)
    })

    authConnection.on('disconnected', event => {
      this.logger.info(`LFA Disconnected!`, event)
      this.libp2pService.emit(Libp2pEvents.AUTH_DISCONNECTED, {
        event,
        connection,
      })
    })

    authConnection.on('joined', async payload => {
      const { team, user } = payload
      const sigChain = this.sigChainService.getActiveChain()
      this.logger.info(
        `${sigChain.localUserContext.user.userId}: Joined team ${team.teamName} (userid: ${user.userId})!`
      )
      if (sigChain.team == null) {
        this.logger.info(
          `${user.userId}: Creating SigChain for user with name ${user.userName} and team name ${team.teamName}`
        )
        sigChain.context = {
          device: (sigChain.context as Auth.InviteeContext).device,
          team,
          user,
        } as Auth.MemberContext
        sigChain.team = team
      }
      this.joinStatus = JoinStatus.JOINED
      this.unblockConnections(this.bufferedConnections)
      this.emit(Libp2pEvents.AUTH_JOINED)
      await this.sigChainService.saveChain(team.teamName)
    })

    authConnection.on('change', payload => {
      this.emit(Libp2pEvents.AUTH_STATE_CHANGED, payload)
    })

    authConnection.on('updated', async head => {
      this.logger.info('Received sync message, team graph updated', head)
      this.emit(Libp2pEvents.AUTH_UPDATED, head)
      const sigChain = this.sigChainService.getActiveChain()
      await this.sigChainService.saveChain(sigChain.team!.teamName)
    })

    // Handle errors from local or remote sources.
    authConnection.on('localError', error => {
      this.emit(Libp2pEvents.AUTH_LOCAL_ERROR, { error, connection })
    })
    authConnection.on('remoteError', error => {
      this.emit(Libp2pEvents.AUTH_REMOTE_ERROR, { error, connection })
    })

    // Store the auth connection and also the underlying libp2p connection
    this.authConnections.set(peerId.toString(), authConnection)
    this.peerConnections.set(peerId.toString(), connection)

    this.logger.info(`Auth connection established with ${peerId.toString()}`)
    authConnection.start()
  }

  private async onPeerDisconnected(peerId: PeerId) {
    if (this.authConnections.has(peerId.toString())) {
      this.logger.warn(`Auth connection with ${peerId.toString()} was disconnected`)
      this.closeAuthConnection(peerId)
    }
  }

  public closeAuthConnection(peerId: PeerId | string) {
    this.logger.info(`Attempting to close auth connection with ${peerId.toString()}`)
    const key = peerId.toString()

    // Remove the stored connection (ephemeral streams are used for each message)
    if (this.peerConnections.has(key)) {
      this.peerConnections.delete(key)
    }

    if (this.authConnections.has(key)) {
      try {
        this.authConnections.get(key)?.stop()
      } catch (e) {
        // do nothing
      }
      this.authConnections.delete(key)
    }
  }
}

export const libp2pAuth = (
  sigChainService: SigChainService,
  libp2pService: Libp2pService
): ((components: Libp2pAuthComponents) => Libp2pAuth) => {
  return (components: Libp2pAuthComponents) => new Libp2pAuth(sigChainService, libp2pService, components)
}
