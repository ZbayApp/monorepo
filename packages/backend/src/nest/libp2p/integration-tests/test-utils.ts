import { Test, TestingModule } from '@nestjs/testing'
import { TestModule } from '../../common/test.module'
import { getLocalLibp2pInstanceParams } from '../../common/utils'
import { Libp2pModule } from '../libp2p.module'
import { SigChainModule } from '../../auth/sigchain.service.module'
import { Libp2pEvents, Libp2pNodeParams } from '../libp2p.types'
import { createLogger } from '../../common/logger'
import { Libp2pService } from '../libp2p.service'

const logger = createLogger('libp2p:test-utils')

export const attachEventListeners = (libp2pService: Libp2pService, timeline: string[], instanceName: string) => {
  // loop over all enum Libp2pEvents and attach event listeners
  for (const event of Object.values(Libp2pEvents)) {
    if (event === Libp2pEvents.AUTH_STATE_CHANGED) {
      libp2pService.on(event, (state: any) => {
        timeline.push(`${event}:${state}`)
      })
    } else {
      libp2pService.on(event, () => {
        timeline.push(`${event}`)
      })
    }
  }
}

export const spawnTestModules = async (number: number) => {
  const modules = []
  for (let i = 0; i < number; i++) {
    const module = await Test.createTestingModule({
      imports: [TestModule, Libp2pModule, SigChainModule],
    }).compile()
    modules.push(module)
  }
  logger.info(`created ${modules.length} test modules`)
  return modules
}

export const spawnLibp2pInstances = async (
  modules: TestingModule[],
  customLibp2pInstanceParams?: Libp2pNodeParams,
  sharePsk: boolean = true
) => {
  logger.info(`creating ${modules.length} libp2p instances`)
  const singlePSK = Libp2pService.generateLibp2pPSK().fullKey
  const libp2pServices = []
  for (let i = 0; i < modules.length; i++) {
    logger.info(`creating libp2p instance ${i}`)
    const libp2pService = await modules[i].resolve(Libp2pService)
    const params = {
      ...(await getLocalLibp2pInstanceParams()),
      ...customLibp2pInstanceParams,
      instanceName: `instance${i}`,
    }
    if (sharePsk) {
      params.psk = singlePSK
    }
    logger.info(`creating libp2p instance with params:`, params)
    await libp2pService.createInstance(params)
    libp2pServices.push(libp2pService)
  }
  logger.info(`created ${libp2pServices.length} libp2p instances`)
  return libp2pServices
}

export const timelinesInclude = (timelines: string[][], event: string): boolean => {
  if (timelines.every(timeline => timeline.includes(event))) {
    return true
  }
  return false
}
