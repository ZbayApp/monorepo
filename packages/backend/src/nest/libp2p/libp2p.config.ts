import { gossipsub } from '@chainsafe/libp2p-gossipsub'
import { identify, identifyPush } from '@libp2p/identify'
import { yamux } from '@chainsafe/libp2p-yamux'
import { tcp } from '@libp2p/tcp'
import { Libp2pOptions, ServiceFactoryMap } from 'libp2p'
import { FaultTolerance } from '@libp2p/interface-transport'
import { bootstrap } from '@libp2p/bootstrap'
import { kadDHT } from '@libp2p/kad-dht'
import { mplex } from '@libp2p/mplex'
import { ping } from '@libp2p/ping'
import { preSharedKey } from '@libp2p/pnet'
import { Libp2pAuth, Libp2pAuthComponents } from './libp2p.auth'
import { MAX_RELAY_DURATION_TIME_LIMIT, WEBSOCKET_CIPHER_SUITE } from './libp2p.const'
import { Libp2pDatastore } from './libp2p.datastore'
import { Libp2pNodeParams } from './libp2p.types'
import { keychain } from '@libp2p/keychain'
import { webSockets as webSocketsOverTor } from '../websocketOverTor'
import * as filters from '@libp2p/websockets/filters'
import { noise, pureJsCrypto } from '@chainsafe/libp2p-noise'
import * as os from 'os'
import { circuitRelayServer, circuitRelayTransport } from '@libp2p/circuit-relay-v2'
import { ServiceMap } from '@libp2p/interface'
import getPort from 'get-port'

export const generateServerLibp2pConfig = async (
  params: Libp2pNodeParams,
  datastore: Libp2pDatastore,
  libp2pAuth: (components: Libp2pAuthComponents) => Libp2pAuth
): Promise<Libp2pOptions> => {
  const transport = webSocketsOverTor({
    filter: filters.all,
    websocket: {
      agent: params.agent,
      handshakeTimeout: 10_000,
      ciphers: WEBSOCKET_CIPHER_SUITE,
      followRedirects: true,
    },
    localAddress: getTcpLocalAddress(params.peerId.peerId.toString(), 3000),
    targetPort: 3000,
    inboundConnectionUpgradeTimeout: 10_000,
    closeOnEnd: false,
    tagSuffix: 'tcp',
  })

  const extraServices: ServiceFactoryMap<ServiceMap> = {
    // @ts-ignore
    circuitRelay: circuitRelayServer({
      maxOutboundStopStreams: 2048,
      maxInboundHopStreams: 2048,
      reservations: {
        maxReservations: 10000,
        defaultDataLimit: BigInt(MAX_RELAY_DURATION_TIME_LIMIT),
        defaultDurationLimit: MAX_RELAY_DURATION_TIME_LIMIT,
      },
    }),
  }

  return generateLibp2pConfig(
    {
      ...params,
      transport: [transport],
      listenAddresses: [...params.listenAddresses, '/p2p-circuit'],
    },
    datastore,
    libp2pAuth,
    extraServices
  )
}

export const generatePeerLibp2pConfig = async (
  params: Libp2pNodeParams,
  datastore: Libp2pDatastore,
  libp2pAuth: (components: Libp2pAuthComponents) => Libp2pAuth
): Promise<Libp2pOptions> => {
  const torTransport = webSocketsOverTor({
    filter: filters.all,
    websocket: {
      agent: params.agent,
      handshakeTimeout: 30_000,
      ciphers: WEBSOCKET_CIPHER_SUITE,
      followRedirects: true,
    },
    localAddress: params.localAddress,
    targetPort: params.targetPort,
    inboundConnectionUpgradeTimeout: 30_000,
    closeOnEnd: false,
    tagSuffix: 'tor',
  })
  const wsPort = await getPort()
  const tcpTransport = webSocketsOverTor({
    filter: filters.all,
    websocket: {
      agent: params.agent,
      handshakeTimeout: 10_000,
      ciphers: WEBSOCKET_CIPHER_SUITE,
      followRedirects: true,
    },
    localAddress: getTcpLocalAddress(params.peerId.peerId.toString(), wsPort),
    targetPort: wsPort,
    inboundConnectionUpgradeTimeout: 10_000,
    closeOnEnd: false,
    tagSuffix: 'tcp',
  })
  const relayTransport = circuitRelayTransport({
    reservationConcurrency: 1024,
  })

  return generateLibp2pConfig(
    {
      ...params,
      transport: [torTransport, tcpTransport, relayTransport],
      listenAddresses: [
        ...params.listenAddresses,
        getTcpListenAddress(params.peerId.peerId.toString(), wsPort),
        '/p2p-circuit',
      ],
    },
    datastore,
    libp2pAuth
  )
}

const generateLibp2pConfig = (
  params: Libp2pNodeParams,
  datastore: Libp2pDatastore,
  libp2pAuth: (components: Libp2pAuthComponents) => Libp2pAuth,
  extraServices?: ServiceFactoryMap<ServiceMap>
): Libp2pOptions => {
  return {
    start: false,
    datastore: datastore.getDatastoreInstance(),
    connectionManager: {
      maxConnections: 20, // TODO: increase?
      dialTimeout: 120_000,
      maxParallelDials: 10,
      inboundUpgradeTimeout: 30_000,
      outboundUpgradeTimeout: 30_000,
      protocolNegotiationTimeout: 10_000,
      maxDialQueueLength: 500,
      reconnectRetries: 25,
    },
    privateKey: params.peerId.privKey,
    addresses: { listen: params.listenAddresses },
    connectionMonitor: {
      // ISLA: we should consider making this true if pings are reliable going forward
      abortConnectionOnPingFailure: false,
      pingInterval: 60_000,
      enabled: true,
    },
    connectionProtector:
      params.useConnectionProtector || params.useConnectionProtector == null
        ? preSharedKey({ psk: params.psk })
        : undefined,
    // @ts-ignore
    peerDiscovery: params.headless
      ? undefined
      : [
          bootstrap({
            list: ['/ip4/127.0.0.1/tcp/3000/ws/p2p/12D3KooWHPcGQbqgwdv386FmFarY27kCiYcYtxgFuBQiEQnuBqGL'],
          }),
        ],
    connectionGater: {
      // Allow private addresses for local testing
      denyDialMultiaddr: async () => false,
    },
    streamMuxers: [
      // mplex({
      //   disconnectThreshold: 20,
      //   maxInboundStreams: 1024,
      //   maxOutboundStreams: 1024,
      //   maxStreamBufferSize: 26214400,
      //   maxUnprocessedMessageQueueSize: 104857600,
      //   maxMsgSize: 10485760,
      //   // @ts-expect-error This is part of the config interface but it isn't typed that way
      //   closeTimeout: 15_000,
      // }),
      yamux({
        maxInboundStreams: 1024,
        maxOutboundStreams: 1024,
        maxMessageSize: 10485760,
      }),
    ],
    // @ts-ignore
    connectionEncrypters: [noise({ crypto: pureJsCrypto, staticNoiseKey: params.peerId.noiseKey })],
    transports: params.transport!,
    transportManager: {
      faultTolerance: FaultTolerance.NO_FATAL,
    },
    services: {
      ...(extraServices ?? {}),
      auth: libp2pAuth,
      ping: ping({ timeout: 30_000 }),
      pubsub: gossipsub({
        // neccessary to run a single peer
        allowPublishToZeroTopicPeers: true,
        fallbackToFloodsub: false,
        emitSelf: true,
        debugName: params.peerId.peerId.toString(),
        doPX: true,
      }),
      identify: identify({ timeout: 30_000, maxInboundStreams: 128, maxOutboundStreams: 128 }),
      identifyPush: identifyPush({ timeout: 30_000, maxInboundStreams: 128, maxOutboundStreams: 128 }),
      keychain: keychain(),
      dht: kadDHT({
        allowQueryWithZeroPeers: false,
        clientMode: false,
        initialQuerySelfInterval: 500,
        providers: {
          cacheSize: 1024,
        },
        maxInboundStreams: 128,
        maxOutboundStreams: 128,
      }),
    },
  }
}

export type IPMap = {
  [interfaceName: string]: string[]
}

export const DEFAULT_NETWORK_INTERFACE = 'en0'

/*
Shamelessly copied from https://stackoverflow.com/a/8440736
*/
export const getIpAddresses = (): IPMap => {
  const interfaces = os.networkInterfaces()
  const results: IPMap = {} // Or just '{}', an empty object

  for (const name in interfaces) {
    for (const net of interfaces[name]!) {
      // Skip over non-IPv4 and internal (i.e. 127.0.0.1) addresses
      // 'IPv4' is in Node <= 17, from 18 it's a number 4 or 6
      const familyV4Value = typeof net.family === 'string' ? 'IPv4' : 4
      if (net.family === familyV4Value && !net.internal) {
        if (!results[name]) {
          results[name] = []
        }
        results[name].push(net.address)
      }
    }
  }

  return results
}

export const getTcpListenAddress = (peerId: string, port: number): string => {
  // const ipAddresses = getIpAddresses()
  // const ipOfInterfaceInUse = ipAddresses[DEFAULT_NETWORK_INTERFACE][0]
  return `/ip4/127.0.0.1/tcp/${port}/ws/p2p/${peerId}`
}

export const getTcpLocalAddress = (peerId: string, port: number): string => {
  // const ipAddresses = getIpAddresses()
  // const ipOfInterfaceInUse = ipAddresses[DEFAULT_NETWORK_INTERFACE][0]
  return `/ip4/127.0.0.1/tcp/${port}/ws/p2p/${peerId}`
}

export const getRelayPeerListenAddress = (peerId: string): string => {
  // const ipAddresses = getIpAddresses()
  // const ipOfInterfaceInUse = ipAddresses[DEFAULT_NETWORK_INTERFACE][0]
  return `/ip4/127.0.0.1/tcp/3000/ws/p2p/12D3KooWHPcGQbqgwdv386FmFarY27kCiYcYtxgFuBQiEQnuBqGL/p2p-circuit/p2p/${peerId}`
}
