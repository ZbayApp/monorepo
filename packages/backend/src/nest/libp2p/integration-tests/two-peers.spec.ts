import { TestingModule } from '@nestjs/testing'
import { Libp2pService } from '../libp2p.service'
import { SigChainService } from '../../auth/sigchain.service'
import { Libp2pEvents } from '../libp2p.types'
import { createLogger } from '../../common/logger'
import { headsAreEqual } from '@localfirst/crdx'
import { spawnLibp2pInstances, spawnTestModules, attachEventListeners, timelinesInclude } from './test-utils'

const logger = createLogger('libp2p:libp2p.auth.spec')

describe('Libp2pAuth', () => {
  const teamName: string = 'team'
  const userA: string = 'instance0'
  const userB: string = 'instance1'
  const eventTimeline: string[] = []
  const eventTimelineA: string[] = []
  const eventTimelineB: string[] = []
  const modules: TestingModule[] = []
  let sigchainServiceA: SigChainService
  let sigchainServiceB: SigChainService
  let libp2pServiceA: Libp2pService
  let libp2pServiceB: Libp2pService

  beforeAll(async () => {
    modules.push(...(await spawnTestModules(2)))
    sigchainServiceA = await modules[0].resolve(SigChainService)
    sigchainServiceB = await modules[1].resolve(SigChainService)
    libp2pServiceA = await modules[0].resolve(Libp2pService)
    libp2pServiceB = await modules[1].resolve(Libp2pService)

    attachEventListeners(libp2pServiceA, eventTimeline, 'A')
    attachEventListeners(libp2pServiceB, eventTimeline, 'B')
    attachEventListeners(libp2pServiceA, eventTimelineA, 'A')
    attachEventListeners(libp2pServiceB, eventTimelineB, 'B')

    // Create chain for instance A
    await sigchainServiceA.createChain(teamName, userA, true)
    expect(sigchainServiceA.activeChainTeamName).toBe(teamName)

    // Create invitation from A -> B
    const inviteResult = await sigchainServiceA.getActiveChain().invites.createLongLivedUserInvite()
    await sigchainServiceB.createChainFromInvite(userB, teamName, inviteResult.seed, true)
  })
  afterEach(() => {
    eventTimeline.length = 0
    eventTimelineA.length = 0
    eventTimelineB.length = 0
  })

  afterAll(async () => {
    for (const module of modules) {
      const libp2pService = await module.resolve(Libp2pService)
      if (libp2pService.libp2pInstance?.status !== 'stopped') {
        await libp2pService.libp2pInstance?.stop()
      }
      await module.close()
    }
  })

  it('create two instances of libp2p', async () => {
    logger.info('Creating libp2p instances')
    await spawnLibp2pInstances(modules)
    logger.info('Created libp2p instances')
    expect(libp2pServiceA.libp2pInstance).toBeDefined()
    expect(libp2pServiceB.libp2pInstance).toBeDefined()
  })

  it('joins with an invitation', async () => {
    logger.info('Initializes chains when peers connect')
    // Wait for chain init event on B
    await new Promise<void>(resolve => {
      libp2pServiceB.once(Libp2pEvents.AUTH_JOINED, () => {
        logger.info('libp2pServiceB initialized chain')
        expect(eventTimeline).toMatchSnapshot('event timeline after joining')
        expect(eventTimelineA).toMatchSnapshot('instanceA event timeline after join')
        expect(eventTimelineB).toMatchSnapshot('instanceB event timeline after join')
        resolve()
      })
      libp2pServiceB.dialPeer(libp2pServiceA.localAddress)
    })

    expect(sigchainServiceA.getActiveChain().team).toBeDefined()
    expect(sigchainServiceB.getActiveChain().team).toBeDefined()
    expect(sigchainServiceA.getActiveChain() !== sigchainServiceB.getActiveChain()).toBeTruthy()
    // at this point B should have added themselves to the chain, but not yet synced that with A
    expect(
      headsAreEqual(
        sigchainServiceA.getActiveChain().team?.graph.head,
        sigchainServiceB.getActiveChain().team?.graph.head
      )
    ).toBeFalsy()
  })
  it('emits connected after syncing', async () => {
    logger.info('Emits connected after syncing')
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('PEER_DISCONNECTED events did not occur within expected time.'))
      }, 70_000)
      const resolveIfMet = async () => {
        if (timelinesInclude([eventTimelineA, eventTimelineB], Libp2pEvents.AUTH_CONNECTED)) {
          logger.info('timelines include expected events!')
          expect(eventTimelineA).toMatchSnapshot('instanceA event timeline after connected')
          expect(eventTimelineB).toMatchSnapshot('instanceB event timeline after connected')
          clearTimeout(timeout)
          resolve()
        }
      }
      libp2pServiceA.on(Libp2pEvents.AUTH_CONNECTED, async () => {
        logger.info('test listener heard libp2pServiceA connected')
        await resolveIfMet()
      })
      libp2pServiceB.on(Libp2pEvents.AUTH_CONNECTED, async () => {
        logger.info('test listener heard libp2pServiceB connected')
        await resolveIfMet()
      })
    })
    // at this point the sync should be complete and the heads should be equal
    expect(
      headsAreEqual(
        sigchainServiceA.getActiveChain().team?.graph.head,
        sigchainServiceB.getActiveChain().team?.graph.head
      )
    ).toBeTruthy()
  })
  it('gracefully disconnects', async () => {
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('PEER_DISCONNECTED events did not occur within expected time.'))
      }, 70_000)
      const resolveIfMet = async () => {
        if (
          eventTimelineA.includes(Libp2pEvents.PEER_DISCONNECTED) &&
          eventTimelineB.includes(Libp2pEvents.PEER_DISCONNECTED)
        ) {
          expect(eventTimelineA).toMatchSnapshot('disconnectEventsA')
          expect(eventTimelineB).toMatchSnapshot('disconnectEventsB')
          expect(eventTimeline).toMatchSnapshot('eventTimeline after disconnection')
          clearTimeout(timeout)
          resolve()
        }
      }
      libp2pServiceA.once(Libp2pEvents.PEER_DISCONNECTED, () => {
        logger.info('test listener heard libp2pServiceA disconnected')
        resolveIfMet()
      })
      libp2pServiceB.once(Libp2pEvents.PEER_DISCONNECTED, () => {
        logger.info('test listener heard libp2pServiceB disconnected')
        resolveIfMet()
      })
      libp2pServiceB.hangUpPeers()
    })
  })
})
