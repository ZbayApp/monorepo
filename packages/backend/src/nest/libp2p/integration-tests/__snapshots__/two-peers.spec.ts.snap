// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Libp2pAuth emits connected after syncing: instanceA event timeline after connected 1`] = `
[
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityAcceptance,verifyingTheirIdentity:done",
  "authStateChanged:negotiating",
  "authStateChanged:synchronizing",
  "authUpdated",
  "authConnected",
  "authStateChanged:connected",
  "authStateChanged:connected",
]
`;

exports[`Libp2pAuth emits connected after syncing: instanceB event timeline after connected 1`] = `
[
  "authStateChanged:negotiating",
  "authStateChanged:synchronizing",
  "authStateChanged:synchronizing",
  "authConnected",
]
`;

exports[`Libp2pAuth gracefully disconnects: disconnectEventsA 1`] = `
[
  "authStateChanged:connected",
  "peerDisconnected",
]
`;

exports[`Libp2pAuth gracefully disconnects: disconnectEventsB 1`] = `
[
  "authDisconnected",
  "authStateChanged:disconnected",
  "peerDisconnected",
]
`;

exports[`Libp2pAuth gracefully disconnects: eventTimeline after disconnection 1`] = `
[
  "authDisconnected",
  "authStateChanged:disconnected",
  "authStateChanged:connected",
  "peerDisconnected",
  "peerDisconnected",
]
`;

exports[`Libp2pAuth joins with an invitation: event timeline after joining 1`] = `
[
  "peerConnected",
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityChallenge,verifyingTheirIdentity:done",
  "authStateChanged:authenticating:awaitingInvitationAcceptance",
  "authJoined",
]
`;

exports[`Libp2pAuth joins with an invitation: instanceA event timeline after join 1`] = `
[
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityChallenge,verifyingTheirIdentity:done",
]
`;

exports[`Libp2pAuth joins with an invitation: instanceB event timeline after join 1`] = `
[
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:awaitingInvitationAcceptance",
  "authJoined",
]
`;
